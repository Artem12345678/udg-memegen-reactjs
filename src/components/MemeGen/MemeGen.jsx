import React, { Component } from "react";
import InputGroup from "react-bootstrap/InputGroup";
import FormControl from "react-bootstrap/FormControl";
import Button from "react-bootstrap/Button";
import ChromePicker from "react-color";

import "./MemeGen.scss";

export default class MemeGen extends Component {
  state = {
    topText: "Start Of Debug",
    bottomText: "End Of Debug",
    fontFamily: "Arial",
    fontSize: 25,
    fontColor: "rgb(255, 0, 255)",
    mimeType: "image/jpeg",
    imageUrl: "https://i.imgflip.com/2cp1.jpg",
    displayColorPicker: false,
    results: [],
    imageName: null
  };

  canvas = React.createRef();

  componentDidMount() {
    this.renderImage();
  }

  componentDidUpdate() {
    this.renderImage();
  }

  get canvasElement() {
    return this.canvas.current;
  }

  get canvasContext() {
    return this.canvasElement.getContext("2d");
  }

  onChangeHandler = event => {
    const { name, value } = event.target;

    this.setState({
      [name]: value
    });
  };

  toggleColorPicker = () => {
    this.setState(prevState => {
      return {
        displayColorPicker: !prevState.displayColorPicker
      };
    });
  };

  onColorChangeHandler = color => {
    const { r, g, b } = color.rgb;

    this.setState({
      fontColor: `rgb(${r}, ${g}, ${b})`
    });
  };

  onFileChangeHandler = event => {
    const file = event.target.files[0];

    const reader = new FileReader();

    reader.onload = () => {
      this.setState({
        imageUrl: reader.result,
        imageName: file.name
      });
    };

    reader.readAsDataURL(file);
  };

  onSaveHandler = event => {
    event.preventDefault();

    const { mimeType, topText, bottomText, results } = this.state;

    let snapshot;

    if (mimeType === "image/gif") {
      // eslint-disable-next-line no-undef
      const gif = new GIF({
        workers: 2,
        quality: 10
      });

      gif.addFrame(this.canvasElement, { delay: 200 });

      gif.on("finished", blob => {
        const reader = new FileReader();

        reader.onload = () => {
          snapshot = reader.result;

          const result = {
            topText,
            bottomText,
            imageUrl: snapshot
          };

          const newResults = [...results, result];

          this.setState({
            results: newResults
          });
        };

        reader.readAsDataURL(blob);
      });

      gif.render();
    } else {
      snapshot = this.canvasElement.toDataURL(mimeType);

      const result = {
        topText,
        bottomText,
        imageUrl: snapshot
      };

      const newResults = [...results, result];

      this.setState({
        results: newResults
      });
    }
  };

  onDeleteHandler = index => {
    const { results } = this.state;

    const newResults = [...results];

    newResults.splice(index, 1);

    this.setState({
      results: newResults
    });
  };

  renderImage = () => {
    const image = new Image();

    const {
      fontSize,
      fontFamily,
      fontColor,
      topText,
      bottomText,
      imageUrl
    } = this.state;

    image.addEventListener("load", () => {
      this.canvasElement.width = image.width;
      this.canvasElement.height = image.height;

      this.canvasContext.drawImage(image, 0, 0);

      this.canvasContext.textAlign = "center";
      this.canvasContext.textBaseline = "middle";
      this.canvasContext.font = `${fontSize}px ${fontFamily}`;
      this.canvasContext.fillStyle = fontColor;
      this.canvasContext.fillText(topText, this.canvasElement.width / 2, 50);
      this.canvasContext.fillText(
        bottomText,
        this.canvasElement.width / 2,
        this.canvasElement.height - 50
      );
    });

    image.setAttribute("crossorigin", "anonymous"); // tainted canvas fix
    image.setAttribute("src", imageUrl);
  };

  render() {
    const {
      fontColor,
      fontFamily,
      fontSize,
      topText,
      bottomText,
      mimeType,
      displayColorPicker,
      imageName,
      results
    } = this.state;

    return (
      <div className="app">
        <div className="googleFontFix">Fix</div>

        <div className="app__toolbar">
          <div className="form-row">
            <InputGroup className="col-lg-4 app__control">
              <InputGroup.Prepend>
                <InputGroup.Text>File</InputGroup.Text>
              </InputGroup.Prepend>
              <div className="custom-file">
                <input
                  id="image"
                  name="file"
                  type="file"
                  accept="image/jpeg,image/png,image/gif"
                  onChange={this.onFileChangeHandler}
                  className="custom-file-input"
                />
                <label className="custom-file-label" htmlFor="image">
                  {imageName || "Select"}
                </label>
              </div>
            </InputGroup>

            <InputGroup className="col-lg-4 app__control">
              <InputGroup.Prepend>
                <InputGroup.Text id="topText">Top Text</InputGroup.Text>
              </InputGroup.Prepend>
              <FormControl
                placeholder="Type text..."
                onChange={this.onChangeHandler}
                name="topText"
                value={topText}
                aria-label="Top Text"
                aria-describedby="topText"
              />
            </InputGroup>

            <InputGroup className="col-lg-4 app__control">
              <InputGroup.Prepend>
                <InputGroup.Text id="bottomText">Bottom Text</InputGroup.Text>
              </InputGroup.Prepend>
              <FormControl
                placeholder="Type text..."
                onChange={this.onChangeHandler}
                name="bottomText"
                value={bottomText}
                aria-label="Bottom Text"
                aria-describedby="bottomText"
              />
            </InputGroup>
          </div>

          <div className="form-row">
            <InputGroup className="col-lg-4 app__control">
              <InputGroup.Prepend>
                <InputGroup.Text id="font">Font</InputGroup.Text>
              </InputGroup.Prepend>
              <FormControl
                className="custom-select"
                as="select"
                name="fontFamily"
                value={fontFamily}
                onChange={this.onChangeHandler}
                aria-label="Font Family"
                aria-describedby="font"
              >
                <option disabled="disabled">Choose</option>
                <option value="Arial">Arial</option>
                <option value="Comic Sans MS">Comic Sans MS</option>
                <option value="Pacifico">Pacifico</option>
              </FormControl>
            </InputGroup>

            <InputGroup className="col-lg-4 app__control">
              <InputGroup.Prepend>
                <InputGroup.Text id="color">Color</InputGroup.Text>
              </InputGroup.Prepend>
              <FormControl
                readOnly
                name="fontColor"
                value={fontColor}
                onClick={this.toggleColorPicker}
                aria-label="Font Color"
                aria-describedby="color"
              />
              {displayColorPicker ? (
                <div className="popover">
                  <div
                    className="cover"
                    onClick={this.toggleColorPicker}
                    role="presentation"
                  />
                  <ChromePicker
                    color={fontColor}
                    onChange={this.onColorChangeHandler}
                  />
                </div>
              ) : null}
            </InputGroup>

            <InputGroup className="col-lg-4 app__control">
              <InputGroup.Prepend>
                <InputGroup.Text id="size">Font Size</InputGroup.Text>
              </InputGroup.Prepend>
              <input
                style={{
                  flexGrow: 1,
                  width: "auto"
                }}
                className="custom-range"
                type="range"
                min="15"
                max="50"
                name="fontSize"
                value={fontSize}
                onChange={this.onChangeHandler}
                aria-label="Font Size"
                aria-describedby="size"
              />
            </InputGroup>
          </div>
        </div>

        <div className="app__canvas">
          <canvas ref={this.canvas} id="canvas" />
        </div>

        <div className="app__save">
          <InputGroup className="col-lg-6 app__control">
            <FormControl
              className="custom-select"
              as="select"
              name="mimeType"
              value={mimeType}
              onChange={this.onChangeHandler}
              aria-label="Mime Type"
            >
              <option disabled="disabled">Choose</option>
              <option value="image/jpeg">JPG</option>
              <option value="image/png">PNG</option>
              <option value="image/gif">GIF</option>
            </FormControl>
            <InputGroup.Append>
              <Button name="save" onClick={this.onSaveHandler} variant="info">
                Save As
              </Button>
            </InputGroup.Append>
          </InputGroup>
        </div>

        <div className="app__results">
          <div className="row">
            {results.map((item, index) => {
              return (
                <div className="col-md-4" key={Math.random()}>
                  <div className="card mb-5">
                    <img
                      className="card-img-top"
                      alt="A custom meme"
                      src={item.imageUrl}
                    />
                    <div className="card-body">
                      <p className="card-text">Top Text: {item.topText}</p>
                      <p className="card-text">
                        Bottom Text: {item.bottomText}
                      </p>
                      <a
                        href={item.imageUrl}
                        download="meme"
                        className="btn btn-success"
                      >
                        Download
                      </a>
                      <button
                        type="button"
                        onClick={() => {
                          this.onDeleteHandler(index);
                        }}
                        className="btn btn-danger"
                      >
                        Delete
                      </button>
                    </div>
                  </div>
                </div>
              );
            })}
          </div>
        </div>
      </div>
    );
  }
}
