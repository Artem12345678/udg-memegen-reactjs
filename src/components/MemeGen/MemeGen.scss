#canvas {
  max-width: 100%;
  max-height: 100%;
}

.googleFontFix {
  font-family: 'Pacifico', cursive;

  position: absolute;

  visibility: hidden;
  clip: rect(0 0 0 0);

  width: 1px;
  height: 1px;
  margin: -1px;
}

.custom-range {
  padding: 0 0.75rem !important;
  background-color: #fff;
  background-clip: padding-box;
  border: 1px solid #ced4da;
  height: calc(1.5em + 0.75rem + 2px) !important;
  border-radius: 0.25rem;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.popover {
  position: absolute;
  z-index: 10;
}

.cover {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.app {
  &__control {
    margin-top: 1.5rem;
  }

  &__canvas {
    display: flex;

    margin-top: 1rem;
    padding: 20px;

    justify-content: center;
    align-items: center;
  }

  &__save {
    display: flex;

    align-items: center;
    justify-content: center;
  }

  &__results {
    margin-top: 2.5rem;
  }
}
