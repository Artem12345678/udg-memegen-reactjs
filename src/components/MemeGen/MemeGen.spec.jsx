import React from "react";
import { shallow } from "enzyme";
import ChromePicker from "react-color";

import MemeGen from "./MemeGen";

describe("MemeGen.jsx", () => {
  let wrapper;

  const data = {
    topText: "Start Of Debug",
    bottomText: "End Of Debug",
    fontFamily: "Arial",
    fontSize: "25",
    fontColor: "rgb(255, 0, 255)",
    mimeType: "image/jpeg",
    imageUrl: "https://i.imgflip.com/2cp1.jpg",
    fontFamilyOptions: [
      { value: null, text: "Choose...", disabled: true },
      { value: "Arial", text: "Arial" },
      { value: "Comic Sans MS", text: "Comic Sans MS" },
      { value: "Pacifico", text: "Pacifico" }
    ],
    mimeTypeOptions: [
      { value: null, text: "Choose...", disabled: true },
      { value: "image/jpeg", text: "JPG" },
      { value: "image/png", text: "PNG" },
      { value: "image/gif", text: "GIF" }
    ],
    results: []
  };

  const onFileChangeHandler = jest.fn();
  const onSaveHandler = jest.fn();
  const onDeleteHandler = jest.fn();
  const renderImage = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<MemeGen />);

    const instance = wrapper.instance();
    instance.onFileChangeHandler = onFileChangeHandler;
    instance.onSaveHandler = onSaveHandler;
    instance.onDeleteHandler = onDeleteHandler;
    instance.renderImage = renderImage;

    wrapper.setState(data);
  });

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("File input element", () => {
    it("exists", () => {
      expect(wrapper.find('[name="file"]').exists()).toBe(true);
    });

    it("triggers onFileChange() method", () => {
      wrapper.find('[name="file"]').simulate("change");

      expect(wrapper.instance().onFileChangeHandler).toBeCalled();
    });
  });

  describe("Top Text input element", () => {
    it("exists", () => {
      expect(wrapper.find('[name="topText"]').exists()).toBe(true);
    });

    it("has right default value", () => {
      expect(wrapper.find('[name="topText"]').prop("value")).toBe(data.topText);
    });

    it("updates corresponding model property", () => {
      const newValue = "Top Text Test";

      const event = {
        target: {
          name: "topText",
          value: newValue
        }
      };

      wrapper.find('[name="topText"]').simulate("change", event);

      expect(wrapper.state("topText")).toBe(newValue);
      expect(wrapper.find('[name="topText"]').prop("value")).toBe(newValue);
    });
  });

  describe("Bottom Text input element", () => {
    it("exists", () => {
      expect(wrapper.find('[name="bottomText"]').exists()).toBe(true);
    });

    it("has right default value", () => {
      expect(wrapper.find('[name="bottomText"]').prop("value")).toBe(
        data.bottomText
      );
    });

    it("updates corresponding model property", () => {
      const newValue = "Bottom Text Test";

      const event = {
        target: {
          name: "bottomText",
          value: newValue
        }
      };

      wrapper.find('[name="bottomText"]').simulate("change", event);

      expect(wrapper.state("bottomText")).toBe(newValue);
      expect(wrapper.find('[name="bottomText"]').prop("value")).toBe(newValue);
    });
  });

  describe("Font Size input element", () => {
    it("exists", () => {
      expect(wrapper.find('[name="fontSize"]').exists()).toBe(true);
    });

    it("has right default value", () => {
      expect(wrapper.find('[name="fontSize"]').prop("value")).toBe(
        data.fontSize
      );
    });

    it("updates corresponding model property", () => {
      const newValue = "50";

      const event = {
        target: {
          name: "fontSize",
          value: newValue
        }
      };

      wrapper.find('[name="fontSize"]').simulate("change", event);

      expect(wrapper.state("fontSize")).toBe(newValue);
      expect(wrapper.find('[name="fontSize"]').prop("value")).toBe(newValue);
    });
  });

  describe("Font Family select element", () => {
    it("exists", () => {
      expect(wrapper.find('[name="fontFamily"]').exists()).toBe(true);
    });

    it("renders the right amount of options", () => {
      expect(wrapper.find('[name="fontFamily"] option').length).toBe(
        data.fontFamilyOptions.length
      );
    });

    it("has right default value", () => {
      expect(wrapper.find('[name="fontFamily"]').prop("value")).toBe(
        data.fontFamily
      );
    });

    it("updates corresponding model property", () => {
      const newValue = "Pacifico";

      const event = {
        target: {
          name: "fontFamily",
          value: newValue
        }
      };

      wrapper.find('[name="fontFamily"]').simulate("change", event);

      expect(wrapper.state("fontFamily")).toBe(newValue);
      expect(wrapper.find('[name="fontFamily"]').prop("value")).toBe(newValue);
    });
  });

  describe("Mime Type select element", () => {
    it("exists", () => {
      expect(wrapper.find('[name="mimeType"]').exists()).toBe(true);
    });

    it("renders the right amount of options", () => {
      expect(wrapper.find('[name="mimeType"] option').length).toBe(
        data.mimeTypeOptions.length
      );
    });

    it("has right default value", () => {
      expect(wrapper.find('[name="mimeType"]').prop("value")).toBe(
        data.mimeType
      );
    });

    it("updates corresponding model property", () => {
      const newValue = "image/gif";

      const event = {
        target: {
          name: "mimeType",
          value: newValue
        }
      };

      wrapper.find('[name="mimeType"]').simulate("change", event);

      expect(wrapper.state("mimeType")).toBe(newValue);
      expect(wrapper.find('[name="mimeType"]').prop("value")).toBe(newValue);
    });
  });

  describe("Colorpicker element", () => {
    it("exists", () => {
      expect(wrapper.find('[name="fontColor"]').exists()).toBe(true);
    });

    it("doesn't show colorpicker itself by default", () => {
      expect(wrapper.find(ChromePicker).exists()).toBe(false);
    });

    it("shows colorpicker itself on click", () => {
      wrapper.find('[name="fontColor"]').simulate("click");

      expect(wrapper.find(ChromePicker).exists()).toBe(true);
    });

    it("updates corresponding model property", () => {
      const newValue = "rgb(255, 255, 255)";

      const event = {
        rgb: {
          r: 255,
          g: 255,
          b: 255
        }
      };

      wrapper.find('[name="fontColor"]').simulate("click");

      wrapper.find(ChromePicker).simulate("change", event);

      expect(wrapper.state("fontColor")).toBe(newValue);
      expect(wrapper.find('[name="fontColor"]').prop("value")).toBe(newValue);
    });
  });

  describe("Save As button element", () => {
    it("exists", () => {
      expect(wrapper.find('[name="save"]').exists()).toBe(true);
    });

    it("triggers onSaveHandler() method", () => {
      wrapper.find('[name="save"]').simulate("click");

      expect(wrapper.instance().onSaveHandler).toBeCalled();
    });
  });

  describe("Results section", () => {
    beforeEach(() => {
      wrapper.setState({
        results: [
          {
            topText: "Top Text Test",
            bottomText: "Bottom Text Test",
            imageUrl: "imageUrl"
          }
        ]
      });
    });

    it("exists", () => {
      expect(wrapper.find(".app__results .row").exists()).toBe(true);
    });

    it("renders the right amount of elements", () => {
      expect(wrapper.find(".app__results .row .col-md-4").length).toBe(
        wrapper.state("results").length
      );
    });

    it("renders right Top Text string", () => {
      expect(
        wrapper
          .find(".app__results .row .col-md-4 p.card-text")
          .at(0)
          .text()
      ).toBe(`Top Text: ${wrapper.state("results")[0].topText}`);
    });

    it("renders right Bottom Text string", () => {
      expect(
        wrapper
          .find(".app__results .row .col-md-4 p.card-text")
          .at(1)
          .text()
      ).toBe(`Bottom Text: ${wrapper.state("results")[0].bottomText}`);
    });

    it("renders right href attribute for the Download link element", () => {
      expect(
        wrapper.find(".app__results .row .col-md-4 .btn-success").prop("href")
      ).toBe(wrapper.state("results")[0].imageUrl);
    });

    describe("Delete button element", () => {
      it("exists", () => {
        expect(
          wrapper.find(".app__results .row .col-md-4 .btn-danger").exists()
        ).toBe(true);
      });

      it("triggers onDeleteHandler() method", () => {
        wrapper
          .find(".app__results .row .col-md-4 .btn-danger")
          .simulate("click");

        expect(wrapper.instance().onSaveHandler).toBeCalled();
      });
    });
  });
});
