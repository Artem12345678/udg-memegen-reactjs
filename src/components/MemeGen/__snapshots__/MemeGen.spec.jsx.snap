// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemeGen.jsx matches snapshot 1`] = `
<div
  className="app"
>
  <div
    className="googleFontFix"
  >
    Fix
  </div>
  <div
    className="app__toolbar"
  >
    <div
      className="form-row"
    >
      <InputGroup
        className="col-lg-4 app__control"
      >
        <InputGroupPrepend>
          <InputGroupText>
            File
          </InputGroupText>
        </InputGroupPrepend>
        <div
          className="custom-file"
        >
          <input
            accept="image/jpeg,image/png,image/gif"
            className="custom-file-input"
            id="image"
            name="file"
            onChange={[MockFunction]}
            type="file"
          />
          <label
            className="custom-file-label"
            htmlFor="image"
          >
            Select
          </label>
        </div>
      </InputGroup>
      <InputGroup
        className="col-lg-4 app__control"
      >
        <InputGroupPrepend>
          <InputGroupText
            id="topText"
          >
            Top Text
          </InputGroupText>
        </InputGroupPrepend>
        <FormControl
          aria-describedby="topText"
          aria-label="Top Text"
          name="topText"
          onChange={[Function]}
          placeholder="Type text..."
          value="Start Of Debug"
        />
      </InputGroup>
      <InputGroup
        className="col-lg-4 app__control"
      >
        <InputGroupPrepend>
          <InputGroupText
            id="bottomText"
          >
            Bottom Text
          </InputGroupText>
        </InputGroupPrepend>
        <FormControl
          aria-describedby="bottomText"
          aria-label="Bottom Text"
          name="bottomText"
          onChange={[Function]}
          placeholder="Type text..."
          value="End Of Debug"
        />
      </InputGroup>
    </div>
    <div
      className="form-row"
    >
      <InputGroup
        className="col-lg-4 app__control"
      >
        <InputGroupPrepend>
          <InputGroupText
            id="font"
          >
            Font
          </InputGroupText>
        </InputGroupPrepend>
        <FormControl
          aria-describedby="font"
          aria-label="Font Family"
          as="select"
          className="custom-select"
          name="fontFamily"
          onChange={[Function]}
          value="Arial"
        >
          <option
            disabled="disabled"
          >
            Choose
          </option>
          <option
            value="Arial"
          >
            Arial
          </option>
          <option
            value="Comic Sans MS"
          >
            Comic Sans MS
          </option>
          <option
            value="Pacifico"
          >
            Pacifico
          </option>
        </FormControl>
      </InputGroup>
      <InputGroup
        className="col-lg-4 app__control"
      >
        <InputGroupPrepend>
          <InputGroupText
            id="color"
          >
            Color
          </InputGroupText>
        </InputGroupPrepend>
        <FormControl
          aria-describedby="color"
          aria-label="Font Color"
          name="fontColor"
          onClick={[Function]}
          readOnly={true}
          value="rgb(255, 0, 255)"
        />
      </InputGroup>
      <InputGroup
        className="col-lg-4 app__control"
      >
        <InputGroupPrepend>
          <InputGroupText
            id="size"
          >
            Font Size
          </InputGroupText>
        </InputGroupPrepend>
        <input
          aria-describedby="size"
          aria-label="Font Size"
          className="custom-range"
          max="50"
          min="15"
          name="fontSize"
          onChange={[Function]}
          style={
            Object {
              "flexGrow": 1,
              "width": "auto",
            }
          }
          type="range"
          value="25"
        />
      </InputGroup>
    </div>
  </div>
  <div
    className="app__canvas"
  >
    <canvas
      id="canvas"
    />
  </div>
  <div
    className="app__save"
  >
    <InputGroup
      className="col-lg-6 app__control"
    >
      <FormControl
        aria-label="Mime Type"
        as="select"
        className="custom-select"
        name="mimeType"
        onChange={[Function]}
        value="image/jpeg"
      >
        <option
          disabled="disabled"
        >
          Choose
        </option>
        <option
          value="image/jpeg"
        >
          JPG
        </option>
        <option
          value="image/png"
        >
          PNG
        </option>
        <option
          value="image/gif"
        >
          GIF
        </option>
      </FormControl>
      <InputGroupAppend>
        <Button
          active={false}
          disabled={false}
          name="save"
          onClick={[MockFunction]}
          type="button"
          variant="info"
        >
          Save As
        </Button>
      </InputGroupAppend>
    </InputGroup>
  </div>
  <div
    className="app__results"
  >
    <div
      className="row"
    />
  </div>
</div>
`;
